### PR TITLE
[DLPACK] Optimize toDLPack Conversion Speed

### LINKED ISSUE #162113: ToDLPack Speed Regression
### 🐛 Describe the bug

Previously in https://github.com/pytorch/pytorch/issues/83069, the toDLPack converter introduces a normalization step that changes the strides to 1 when shape[i] == 1

This step, however, calls as_strided during toDLPack, and can slow down the toDLPack about 3x. This causes PyTorch's DLPack conversion to be around 0.6 us overhead per call from the < 0.2us.

Most of the framework handles toDLPack conversion around 0.1 - 0.2 us level, which helps to have good normal eager mode conversion with torch

cc @msaroufim @jerryzh168

### PR DESCRIPTION
**Pull request by tqchen**
Previously in gh-83069, the toDLPack converter introduces a normalization step that changes the strides to 1 when shape[i] == 1

This step, however, calls as_strided during toDLPack, and can slow down the toDLPack about 3x. This causes PyTorch's DLPack conversion to be around 0.6 us overhead per call from the < 0.2us.

This PR updates the logic by adding a need_normalize_strides check, to first confirm if the strides normalization is necessary. In most common cases, when the tensor is continguous, such normalization is not necessary.

We confirmed that having this additional step would recover the speed of toDLPack to below 0.2us and can help significantly speedup eager mode integration of DLPack with PyTorch.

If we detect that there is normalization needs, the older path will be invoked.

Fixes #162113

### CODE REVIEW

**Comment by tqchen** [comment]
Benchmark, on AMD Ryzen:

```
torch.utils.dlpack.to_dlpack[old]             6.162643432617187e-07 sec/call. 
to_dlpack[new]                                          1.8970966339111327e-07 sec/call
```

**Review by msaroufim — APPROVED** [review]
@pytorchbot merge
